routing/router.go

```diff
@@ -1639,98 +1719,288 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 
 	timeoutChan := time.After(payAttemptTimeout)
 
+	// Get the current state of this payment, such that we can resume where
+	// we left off.
+	paymentHash := payment.PaymentHash
+	state, err := r.paymentStates.fetchPaymentState(paymentHash)
+	if err != nil {
+		return [32]byte{}, nil, err
+	}
+
 	// We'll continue until either our payment succeeds, or we encounter a
 	// critical error during path finding.
-	var lastError error
 	for {
-		// Before we attempt this next payment, we'll check to see if
-		// either we've gone past the payment attempt timeout, or the
-		// router is exiting. In either case, we'll stop this payment
-		// attempt short.
-		select {
-		case <-timeoutChan:
-			errStr := fmt.Sprintf("payment attempt not completed "+
-				"before timeout of %v", payAttemptTimeout)
+		switch state {
```

Comment:
I took a stab at it here: https://github.com/lightningnetwork/lnd/pull/2761/commits/ce35b9996ad0bb1be80ce5ca634c4341ba8e7b1d

You think it would be beneficial to split it case into its own method?

Reply:
I indeed was thinking about splitting the cases.

With all the parameters it feels like we are moving towards a payment struct that contains them and the funcs that operate on it.

I am not sure about that `isTerminal`. Only used once and may not make it clearer. Could also do this with a intermediate var `isTerminal := ... ; if isTerminal {}`